File: Tree/bst.cpp
#include <bits/stdc++.h>

using namespace std;

class Node
{

public:
    int data;
    Node *left, *right;

    Node()
    {
        left = right = NULL;
    }

    Node(int data)
    {
        this->data = data;
        left = right = NULL;
    }
};

void print(Node *root)
{
    if (root == NULL)
        return;

    print(root->left);
    cout << root->data << " ";
    print(root->right);
}

int isBST(Node *root)
{
    if (root == NULL)
        return 1;

    static Node *prev = NULL;

    if (!isBST(root->left))
        return 0;

    if (prev != NULL && root->data <= prev->data)
        return 0;

    prev = root;

    return isBST(root->right);
}

Node *search(Node *root, int key)
{
    if (root == NULL)
        return NULL;

    if (root->data == key)
        return root;

    else if (key < root->data)
        return search(root->left, key);
    else
        return search(root->right, key);
}

Node *searchForInsert(Node *root, int data)
{
    if (root == NULL)
        return NULL;

    static Node *prev = NULL;
    prev = root;

    if (root->data == data)
    {
        prev = NULL;
        return prev;
    }

    if (data < root->data)
        searchForInsert(root->left, data);

    else
        searchForInsert(root->right, data);

    return prev;
}

void insert(Node *&root, int data)
{
    Node *newNode = new Node(data);

    if (root == NULL)
    {
        root = newNode;
        return;
    }

    Node *prev = searchForInsert(root, data);

    if (prev != NULL)
    {
        if (data < prev->data)
        {
            prev->left = newNode;
        }
        else
        {
            prev->right = newNode;
        }
    }
}

int main()
{

    Node *root = NULL;

    // Node *n1 = new Node(5);
    // Node *n2 = new Node(3);
    // Node *n3 = new Node(6);
    // Node *n4 = new Node(1);
    // Node *n5 = new Node(4);

    // n1->left = n2;
    // n1->right = n3;

    // n2->left = n4;
    // n2->right = n5;

    int n;
    cin >> n;

    for (int i = 0; i < n; i++)
    {
        int data;
        cin >> data;
        insert(root, data);
    }

    print(root);

    cout << endl;

    return 0;
}

File: Python/Tree/Practice/bst.py
class Node:
    def __init__(self, data, left=None, right=None):
        self.data = data
        self.left = left
        self.right = right


class BinarySearchTree:
    def __init__(self, root=None):
        self.root = root

    def _insert(self, data, root):
        if not root:
            return Node(data)

        if data < root.data:
            root.left = self._insert(data, root.left)
        else:
            root.right = self._insert(data, root.right)

        return root

    def insert(self, data):
        self.root = self._insert(data, self.root)

    def _preOrder(self, root):
        if root:
            print(root.data, end=' ')
            self._preOrder(root.left)
            self._preOrder(root.right)

    def preOrder(self):
        self._preOrder(self.root)
        print()

    def _inOrder(self, root):
        if root:
            self._inOrder(root.left)
            print(root.data, end=' ')
            self._inOrder(root.right)

    def inOrder(self):
        self._inOrder(self.root)
        print()

    def _postOrder(self, root):
        if root:
            self._postOrder(root.left)
            self._postOrder(root.right)
            print(root.data, end=' ')

    def postOrder(self):
        self._postOrder(self.root)
        print()

    def _search(self, root, data, path=None):
        if not root:
            return None

        if path is None:
            path = []

        if data == root.data:
            path.append(root.data)
            return root

        if data < root.data:
            path.append(root.data)
            return self._search(root.left, data, path)
        else:
            path.append(root.data)
            return self._search(root.right, data, path)

    def search(self, data):
        path = []
        return self._search(self.root, data, path), path

    def _delete(self, root, key):
        if not root:
            return root

        if key < root.data:
            root.left = self._delete(root.left, key)
        elif key > root.data:
            root.right = self._


if __name__ == '__main__':
    bst = BinarySearchTree()

    bst.insert(8)
    bst.insert(3)
    bst.insert(10)
    bst.insert(1)
    bst.insert(6)
    bst.insert(4)
    bst.insert(7)
    bst.insert(14)
    bst.insert(13)

    bst.inOrder()

    bst.delete(6)

    bst.inOrder()
